py: Implement PEP 526, syntax for variable annotations

dpgeorge opened the pull request:
This addition to the grammar was introduced in Python 3.6 (see issue #2415).  It allows annotating the type of a varilable, like:
```python
x:int = 123
s:str
```
The implementation here is quite simple and just ignores the annotation (the `int` and `str` bits above).  The reason to implement this is to allow Python 3.6+ code that uses this feature to compile under MicroPython without change.

In the future viper could use this syntax as a way to give types to variables (which is currently done in a bit of an ad-hoc way, eg `x = int(123)`).  And this syntax could potentially be used in the inline assembler to define labels in an way that's easier to read.

dpgeorge commented:
See https://www.python.org/dev/peps/pep-0526/ for the PEP.

Code size change for this PR is:
```
   bare-arm:   +44 +0.067% 
minimal x86:   +80 +0.055% 
   unix x64:   +40 +0.008% 
unix nanbox:   +48 +0.011% 
      stm32:   +44 +0.011% PYBV10
     cc3200:   +56 +0.031% 
    esp8266:   +68 +0.010% GENERIC
      esp32:   +80 +0.006% GENERIC[incl -16(data)]
        nrf:   +52 +0.036% pca10040
       samd:   +60 +0.059% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

tve commented:
What prompted this implementation? I guess it's "nice to have" but last I looked there are python 3.5 features that are not yet implemented?

dpgeorge commented:
> What prompted this implementation?

It was on my to-do list for a long time... at some point I was trying to run code that used this syntax and it would not run.

This is not high-priority but is one of the things that eventually needs to be done to get full Python 3.6 syntax compatibility.

kevinkk525 commented:
I have been waiting for this feature for a long time. Thanks!

@tve with bigger projects I found it very useful to use variable annotation. It makes writing code easier and more robust (at least in my case). Also easier to get back into the code after some time.
Until now, I had to run a script "strip_hints" over all project files before compiling/copying them because upy didn't support variable annotations. 
So for me this was a feature that can't wait until all requested python 3.5 features have been built in, because this is the way people could/should write programs now (with python >3.6).

dpgeorge committed:
py/grammar.h: Consolidate duplicate sub-rules for :test and =test.

dpgeorge committed:
py/compile: Implement PEP 526, syntax for variable annotations.

This addition to the grammar was introduced in Python 3.6.  It allows
annotating the type of a varilable, like:

    x: int = 123
    s: str

The implementation in this commit is quite simple and just ignores the
annotation (the int and str bits above).  The reason to implement this is
to allow Python 3.6+ code that uses this feature to compile under
MicroPython without change, and for users to use type checkers.

In the future viper could use this syntax as a way to give types to
variables, which is currently done in a bit of an ad-hoc way, eg
x = int(123).  And this syntax could potentially be used in the inline
assembler to define labels in an way that's easier to read.

dpgeorge committed:
tests/basics: Add tests for variable annotations.

beyonlo commented:
@kevinkk525 Do you have any example to share using this syntax for variable annotations on MicroPython to have a robust code?

I would like to check, for example, when the method receive a wrong argument type, it return a error for that, or some like that. Is this the main power of this feature?

Thank you.

peterhinch commented:
Annotations are not checked by the compiler, they are for information only. There may be static analysis tools that can do this.

mattytrentini commented:
@beyonlo [mypy](http://mypy-lang.org/) is a commonly used tool for this. Consider the following:

```python
# th.py
def take_int(x: int):
    print(x)

take_int(10)
take_int("abcd")
```

(Micro)Python runs this without error, not checking types:

```bash
❯ python th.py
10
abcd
```

But running mypy quickly identifies the issue:
```bash
❯ mypy .
th.py:5: error: Argument 1 to "take_int" has incompatible type "str"; expected "int"
Found 1 error in 1 file (checked 1 source file)
```

beyonlo commented:
@mattytrentini 

Understood. But in this case I need to have all methods/functions available to be called by `mypy` using like as a `test.py` - that will be pass argument test for all methods, right? However some methods can't be called directly because are internally, I mean, just some methods can to call another methods. Do you have a tip to how to figure out that?

Thank you.

ZanderBrown commented:
It's _static_ analysis — it's not running your code
